Delegate slf4j in HostClassLoader to system class loader

hubertp opened the pull request:
### Pull Request Description

That way in Enso and Java code any usage of slf4j will use the same configuration as the rest of the process that started it.

With
```diff
diff --git a/build.sbt b/build.sbt
index df8e5fdfd4..e0a1b4ccc5 100644
--- a/build.sbt
+++ b/build.sbt
@@ -4484,7 +4484,8 @@ lazy val `std-base` = project
     libraryDependencies ++= Seq(
       "org.graalvm.polyglot"       % "polyglot"                % graalMavenPackagesVersion,
       "org.netbeans.api"           % "org-openide-util-lookup" % netbeansApiVersion % "provided",
-      "com.fasterxml.jackson.core" % "jackson-databind"        % jacksonVersion
+      "com.fasterxml.jackson.core" % "jackson-databind"        % jacksonVersion,
+      "org.slf4j"                  % "slf4j-api"               % JPMSUtils.slf4jVersion,
     ),
     Compile / packageBin := Def.task {
       val result = (Compile / packageBin).value
diff --git a/distribution/lib/Standard/Base/0.0.0-dev/src/Random.enso b/distribution/lib/Standard/Base/0.0.0-dev/src/Random.enso
index c1daaa75b1..7d2ad76a6e 100644
--- a/distribution/lib/Standard/Base/0.0.0-dev/src/Random.enso
+++ b/distribution/lib/Standard/Base/0.0.0-dev/src/Random.enso
@@ -16,6 +16,7 @@ import project.System
 from project.Data.Boolean import Boolean, False, True
 from project.Data.Numbers import Float, Integer
 from project.Data.Range.Extensions import all
+from project.Logging import all
 
 polyglot java import java.lang.Integer as Java_Integer
 polyglot java import java.util.Random as Java_Random
@@ -293,6 +294,7 @@ type Random_Generator
     ## PRIVATE
     boolean : Boolean
     boolean self =
+        self.log_message "Foo!" Log_Level.Warning
         self.java_random.nextBoolean
 
     ## PRIVATE
diff --git a/std-bits/base/src/main/java/org/enso/base/random/Random_Utils.java b/std-bits/base/src/main/java/org/enso/base/random/Random_Utils.java
index 9d11d23bdd..11e82b240f 100644
--- a/std-bits/base/src/main/java/org/enso/base/random/Random_Utils.java
+++ b/std-bits/base/src/main/java/org/enso/base/random/Random_Utils.java
@@ -2,10 +2,15 @@ package org.enso.base.random;
 
 import java.util.Arrays;
 import java.util.Random;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class Random_Utils {
+  private static Logger logger = LoggerFactory.getLogger(Random_Utils.class);
+
   /** Default `seed` used to initialize new instances of `Random` */
   public static long getDefaultSeed() {
+    logger.info("Bar!");
     return java.lang.System.nanoTime();
   }
```

and project containing an actual call
```
node1 = Random.new_generator Nothing
node2 = node2.boolean
```
it would print the expected log messages in the format matching the rest of logs:
```
[INFO] [2024-11-26T17:00:31+01:00] [org.enso.base.random.Random_Utils] Bar!
[WARN] [2024-11-26T17:00:31+01:00] [Standard.Base.Random.Random_Generator] Foo!
```

Previously no logs would be present or they would misformatted.

Closes #11641 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Delegate slf4j in HostClassLoader to system class loader

That way in Enso and Java code any usage of slf4j will use the same
configuration as the rest of the process that started it.
Tricky to test, will post an example in GH PR.

hubertp commented:
No tests are present atm as we currently have no logs in stdlib components or std-bits. I'd prefer not to add any artificially just for the sake of it. With time, that may obviously change.

I modified `micro-distribution` to log something (in a similar spirit to [StdLibLogsTest](https://github.com/enso-org/enso/blob/develop/engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/StdLibLogsTest.java)) but that seems to work with and without this change, so it seemed useless as a test.

Happy to consider suggestions.

4e6 commented:
To test the logging, you can create runtime context that gathers the log output: https://github.com/enso-org/enso/blob/9ddd8417b243f5e6fba1b6a8682224be7d705b8d/engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala#L70 Then run the program and check that the output contains the desired line.

Akirathan commented:
What about a test like this?
```diff
diff --git a/engine/runtime-integration-tests/src/test/java/org/enso/example/Logging_Test_Utils.java b/engine/runtime-integration-tests/src/test/java/org/enso/example/Logging_Test_Utils.java
index 208a32c2f0..8f8d78afa5 100644
--- a/engine/runtime-integration-tests/src/test/java/org/enso/example/Logging_Test_Utils.java
+++ b/engine/runtime-integration-tests/src/test/java/org/enso/example/Logging_Test_Utils.java
@@ -1,4 +1,4 @@
-package org.enso.base_test_helpers;
+package org.enso.example;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -6,7 +6,8 @@ import org.slf4j.LoggerFactory;
 public class Logging_Test_Utils {
   private static Logger logger = LoggerFactory.getLogger(Logging_Test_Utils.class);
 
-  public static void logSomething() {
-    logger.info("Logging something");
+  public static boolean logSomething() {
+    logger.warn("Logging something");
+    return true;
   }
 }
diff --git a/engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/SLF4JLoggingTest.java b/engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/SLF4JLoggingTest.java
index ba85c19194..510fed0ff9 100644
--- a/engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/SLF4JLoggingTest.java
+++ b/engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/SLF4JLoggingTest.java
@@ -1,4 +1,54 @@
 package org.enso.interpreter.test;
 
+import static org.junit.Assert.assertTrue;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import org.enso.common.RuntimeOptions;
+import org.enso.test.utils.ContextUtils;
+import org.graalvm.polyglot.Context;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+
+
 public class SLF4JLoggingTest {
+
+  private static Context ctx;
+  private static ByteArrayOutputStream out;
+
+  @BeforeClass
+  public static void initEnsoContext() {
+    out = new ByteArrayOutputStream();
+    ctx = ContextUtils.defaultContextBuilder()
+        .out(out)
+        .err(out)
+        .logHandler(out)
+        .option(RuntimeOptions.LOG_LEVEL, "INFO")
+        .build();
+  }
+
+  @AfterClass
+  public static void disposeContext() throws IOException {
+    out.close();
+    out = null;
+    ctx.close();
+    ctx = null;
+  }
+
+  @Test
+  public void slf4jLoggerUsedViaPolyglot_ShouldLogSomething() {
+    var src = """
+        polyglot java import org.enso.example.Logging_Test_Utils
+        
+        main =
+            Logging_Test_Utils.logSomething
+        """;
+    var res = ContextUtils.evalModule(ctx, src);
+    assertTrue(res.isBoolean());
+    assertThat(out.toString(), containsString("Logging"));
+  }
 }
```

I tried to apply that patch on your PR, but that test fails for me. Any ideas how to make it work? I believe that should test the functionality. No?

Akirathan commented:
> To test the logging, you can create runtime context that gathers the log output: 

@4e6 That is not sufficient. You need to call Enso code (from Java code), that calls Java code, that initializes SLF4J logger.

hubertp committed:
Semi-working test

hubertp commented:
> I tried to apply that patch on your PR, but that test fails for me. Any ideas how to make it work? I believe that should test the functionality. No?

Sadly no, it's the same issue I had with micro-distribution. It also tests only logging in Java not in Enso.
I added it anyway but you can see that if you revert my patch `HostClassLoader` it still passes, as it seems that tests don't reflect exactly the problem with separate classloaders (different modules?).

hubertp committed:
Make sure memory appender starts clean

Appender is re-used among many tests, collecting all logs. This is
problematic when we want to test the contents of logs.

JaroslavTulach commented:
Testing `enso` launcher from inside of Enso and checking the output of the process is done for example in [startup benchmarks](https://github.com/enso-org/enso/blob/develop/test/Benchmarks/src/Startup/Startup.enso#L30).

hubertp commented:
> Testing `enso` launcher from inside of Enso and checking the output of the process is done for example in [startup benchmarks](https://github.com/enso-org/enso/blob/develop/test/Benchmarks/src/Startup/Startup.enso#L30).

That would work if we reported any logs inside stdlib. Currently we don't (catch-22). That will obviously change over time and I would be happy to add it at that point.

Akirathan commented:
> That would work if we reported any logs inside stdlib. Currently we don't (catch-22). That will obviously change over time and I would be happy to add it at that point.

What about https://github.com/enso-org/enso/compare/wip/jd/table-viz-perf @jdunkerley . Does the logging that you are doing on your branch work for you now?

<blockquote><img src="https://opengraph.githubassets.com/2edd2979bfb7c18543c4ce3d7aef5bdf367920b1e2dcaa22c7184bc0a343951b/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Comparing develop...wip/jd/table-viz-perf · enso-org/enso</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Comparing develop...wip/jd/table-viz-perf · enso-org/enso</div></blockquote>

hubertp commented:
> What about https://github.com/enso-org/enso/compare/wip/jd/table-viz-perf @jdunkerley . Does the logging that you are doing on your branch work for you now?

That _semi-worked_ because I told him how to hack the setup. The (uncommitted) hack is no longer necessary.
